augur/ancestral.py

```diff
@@ -191,36 +247,46 @@ def run(args):
     # explicitly or by default) and the user has not explicitly requested that
     # we keep them.
     infer_ambiguous = args.infer_ambiguous and not args.keep_ambiguous
-
-    tt = ancestral_sequence_inference(tree=T, aln=aln, ref=ref, marginal=args.inference,
-                                      fill_overhangs = not(args.keep_overhangs),
-                                      infer_tips = infer_ambiguous)
-
-    character_map = {}
-    for x in tt.gtr.profile_map:
-        if tt.gtr.profile_map[x].sum()==tt.gtr.n_states:
-            # TreeTime treats all characters that are not valid IUPAC nucleotide chars as fully ambiguous
-            # To clean up auspice output, we map all those to 'N'
-            character_map[x] = 'N'
-        else:
-            character_map[x] = x
-
-    anc_seqs.update(collect_mutations_and_sequences(tt, full_sequences=not is_vcf,
-                          infer_tips=infer_ambiguous, character_map=character_map))
-    # add reference sequence to json structure. This is the sequence with
-    # respect to which mutations on the tree are defined.
-    if is_vcf:
-        anc_seqs['reference'] = {"nuc":compress_seq['reference']}
-    else:
-        root_seq = tt.sequence(T.root, as_string=False)
-        if anc_seqs.get("mask") is not None:
-            root_seq[anc_seqs['mask']] = tt.gtr.ambiguous
-        anc_seqs['reference'] = {"nuc": ''.join(root_seq)}
+    full_sequences = not is_vcf
+    nuc_result = run_ancestral(T, aln, root_sequence=str(ref.seq) if ref else None, is_vcf=is_vcf, fill_overhangs=not args.keep_overhangs,
+                               full_sequences=full_sequences, marginal=args.inference, infer_ambiguous=infer_ambiguous, alphabet='nuc')
+    anc_seqs = nuc_result['mutations']
+    anc_seqs['reference'] = {'nuc': nuc_result['root_seq']}
 
     if anc_seqs.get("mask") is not None:
         anc_seqs["mask"] = "".join(['1' if x else '0' for x in anc_seqs["mask"]])
 
-    anc_seqs['annotations'] = {'nuc': {'start': 1, 'end': len(anc_seqs['reference']['nuc']), 'strand': '+'}}
+    anc_seqs['annotations'] = {'nuc': {'start': 1, 'end': len(anc_seqs['reference']['nuc']),
+                                       'strand': '+', 'type': 'source'}}
+
+    if not is_vcf and args.genes:
+        from .utils import load_features
+        ## load features; only requested features if genes given
+        features = load_features(args.annotation, args.genes)
```

Comment:
I'm not sure whether it's fair to lump it into this PR but [our annotations parsing needs a lot of work](https://github.com/nextstrain/augur/issues/953). 

Testing this PR (for HepB) with `NC_003977.gb` (straight from GenBank):

```
File "***/augur/augur/ancestral.py", line 272, in run
feat = features[gene]
KeyError: 'X'
```

Testing with the nextclade GFF version of the same annotation, `genemap.gff`:

```
Couldn't find gene X in GFF or GenBank file
Couldn't find gene pre-capsid in GFF or GenBank file
Couldn't find gene capsid in GFF or GenBank file
Couldn't find gene pol in GFF or GenBank file
Couldn't find gene envL in GFF or GenBank file
Couldn't find gene envM in GFF or GenBank file
Couldn't find gene envS in GFF or GenBank file
Read in 0 features from reference sequence file
Processing gene: X
Traceback (most recent call last):
  File "/Users/enigma/projects/nextstrain/augur/augur/__init__.py", line 66, in run
    return args.__command__.run(args)
  File "/Users/enigma/projects/nextstrain/augur/augur/ancestral.py", line 272, in run
    feat = features[gene]
KeyError: 'X'
```

Reply:
The following GFF file works. However as this is intended to be run in conjunction with Nextclade we should really sort the parsing out!

```console
##gff-version 3
#!gff-spec-version 1.21
#!processor NCBI annotwriter
##sequence-region NC_003977.2 1 3182
##species https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=10407
NC_003977.2	RefSeq	region	1	3182	.	+	.	ID=NC_003977.2:1..3182;Dbxref=taxon:10407;Is_circular=true;gbkey=Src;genome=genomic;mol_type=genomic DNA;strain=ayw
NC_003977.2	RefSeq	gene	1376	1840	.	+	0	gene=X;
NC_003977.2	RefSeq	gene	1816	2454	.	+	0	gene=pre-capsid;
NC_003977.2	RefSeq	gene	1903	2454	.	+	0	gene=capsid;
NC_003977.2	RefSeq	gene	2309	4807	.	+	0	gene=pol;
NC_003977.2	RefSeq	gene	2850	4019	.	+	0	gene=envL;
NC_003977.2	RefSeq	gene	3174	4019	.	+	0	gene=envM;
NC_003977.2	RefSeq	gene	157	837	.	+	0	gene=envS;
```

Reply:
yeah, this is a bit tricky. Our parsing is first limiting itself to `source` and `gene`, and then ignores "subfeatures". The GFF parser parses CDS that are children of a gene into `feat.subfeatures`. So yes, this is something we need to tackle. 

What we want `load_features` to do is to return a source/landmark feature and then all the bits that need translating, probably best as `type='CDS'` rather than `type='gene'`.

Reply:
I added a simpler parser for the nextclade json here: 
https://github.com/nextstrain/augur/pull/1263